### ISSUE TITLE
Absolute target_lens of 1 gets casted to 0 in ctc_loss

### ISSUE THREAD

**Issue by gebauertnt** [speechbrain/speechbrain#1129]
When converting the relative target_lens into absolute target_lens within the ctc_loss the special case of an absolute target length of 1 gets casted to 0. This results into an infinite loss, when reduction "batch" is applied. For us just adding a very small number to the relative length fixes it for now, but a catch within the ctc_loss function would probably be more convenient.

**Comment by mravanelli** [comment]
Hi @gebauertnt,
thank you for reporting that. I guess you are referring to this line here: 
https://github.com/speechbrain/speechbrain/blob/develop/speechbrain/nnet/losses.py#L237

I'm not sure where the 0 can come out. For instance, take a look at the following code snippet:

```
>>> targets = torch.rand(3, 1)
# Relative lengths
>>> target_lens = torch.Tensor([1.0, 1.0, 1.0])
# From relative to absolute lengths
>>> target_lens = (target_lens * targets.shape[1]).int()
>>> target_lens
tensor([1, 1, 1], dtype=torch.int32)
```

Here it looks like the absolute length is 1 as expected. Could you provide a counterexample that is affected by the issue you mentioned?

**Comment by mravanelli** [comment]
My impression is that you might have manipulated the target tensor (e.g, removing some labels) without modifying the corresponding length vector. Is that the case in your code?

**Comment by gebauertnt** [comment]
Hi @mravanelli,

unfortunately, this happens due to float32. We do not manipulate the target tensor, but have uncommon length differences. We definitely should do some sorting, but still this should not happen. Below a short sample

```
>>> (41.0*torch.Tensor([1.0/41.0, 41.0/41.0])).int()
tensor([ 0, 41], dtype=torch.int32)
```
 Increasing to a length of 2 and 82 also results in a similar mistake. Smaller differences will work just fine.

**Comment by mravanelli** [comment]
Hi @gebauertnt,
I'm able to replicate the issue. This is something weird and I don't really understand why this is happening. It mainly looks a bug in torch that happens only with some specific numbers.  For instance:

**Correct (most of the times):**
```
>>> (15.0 * torch.tensor([1.0/15.0], dtype=torch.float32)).int()
tensor([1], dtype=torch.int32)
```
```
>>> (42.0 * torch.tensor([1.0/42.0], dtype=torch.float32)).int()
tensor([1], dtype=torch.int32)
```

```
>>> (4540.0 * torch.tensor([1.0/4540.0], dtype=torch.float32)).int()
tensor([1], dtype=torch.int32)
```

**Wrong:**

```
>>> (41.0 * torch.tensor([1.0/41.0], dtype=torch.float32)).int()
tensor([0], dtype=torch.int32)
```
```
>>> (82.0 * torch.tensor([1.0/82.0], dtype=torch.float32)).int()
tensor([0], dtype=torch.int32)
```

One way to fix it is to convert the tensor to float64:

```
>>> (41.0 * torch.tensor([1.0/41.0], dtype=torch.float64)).int()
tensor([1], dtype=torch.int32)
```

Do you confirm it? However, I don't really know the reason.

**Comment by mravanelli** [comment]
I had a chat with the pytorch team about this issue:

https://github.com/pytorch/pytorch/issues/68561

Basically, we should replace `.int()` with `.round()`. `Int()` rounds towards "0" and in some cases, due to imprecision of float32, we can have numbers like "0.999999999991" that are assigned to a 0 rather than to a 1. 

Tomorrow I will double check all the recipes potentially affected by that and I will create a PR with the fix.

**Comment by gebauertnt** [comment]
Thank you for the quick fix!

**Comment by mravanelli** [comment]
This is solved in the current dev branch. Thank you @gebauertnt!

**Comment by boeddeker** [comment]
@mravanelli Can you explain, why speechbrain uses relative lengths? (I have/had the same issue as gebauertnt)
I don't get it, why people use floating point numbers, while integers could be used.
Integers have no rounding issues.

For example, in Kaldi they use seconds instead of samples to mark the start and end time of an utterance.
I had many troubles with that, because I had to replicate the rounding errors.
I was hoping, that speechbrain avoids such unnecessary issues.

**Comment by mravanelli** [comment]
Hi @boeddeker,
the rounding issue should be solved in #1144. 

We use relative lengths for a couple of reasons. One is that models often internally change their time resolution using layers such as convolutions with stride factors > 1 or pooling. Other operations like feature extraction use sliding windows that change the time resolution too. With the relative length approach, you can still compute (not always exactly but with reasonable precision) the number of valid output time steps even if the model internally uses these operations (you just have to multiply by the _batch dimensional * relative length_ and round it). The alternative is to always return the absolute length after every basic operation (e.g., after each convolutional layer, pooling, etc.). This would be more transparent,  but a bit annoying because you need to propagate the actual length through all the computational chains.  The absolute length thus introduces boilerplate code and makes it hard to use popular functionalities such as torch.sequential (that only expects one input and one output). 

However, I think we will dive into this solution for the next major version of SpeechBrain. For instance, in #933 we already implemented a mask-capable sequential container that can implement this (the big problem is that this construct is not jitable and we have to still find a way to have it jitable).

**Comment by boeddeker** [comment]
Ah, ok, thank you. I initially used a PyPi version of speechbrain and got an old version without the fix.
Now I use the github version with the fix.

I agree, that it is easier to use the relative length, when the network changes the time resolution.
But on the other side, it can also introduce strange errors, because some operations have only a roughly linear effect on the duration ratio, e.g. pooling may use padding at the end, so the relative length may increase with each layer.
Or, as in this issue, it is frequently happening, that someone forgets the call of round.
I hope, that the durations don't get too large in speechbrain. At some point, the issue changes, that you cannot represent a particular number (i.e. float 32 has a resolution of 1e-06, so values around 1 000 000 or higher are difficult).

Good to know, that you are working on an exact solution.

Thanks again for your explanation.

**Comment by mravanelli** [comment]
Yes, we will likely switch to a mask-based approach for the next major
version.
